### PR TITLE
DM-38632: Add NormalizedCompensatedFluxTask and update aperture corrections.

### PR DESCRIPTION
**Pull request by erykoff**
Along with the new NormalizedCompensatedFluxTask (to normalize compensated tophats and gaussians) this adds a new "median shift" to the aperture corrections that helps reduce scatter in wide field images.

### CODE REVIEW

File: tests/test_normalizedCalibrationFlux.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        # aperture flux offset is 0.0
+        ratio = np.mean(catalog["slot_CalibFlux_instFlux"]/catalog[self.ap_name + "_instFlux"])
+        self.assertFloatsAlmostEqual(ratio, 1.0, rtol=1e-10)
+
+        # The subset that was used should always have a 1.0 ratio.
+        used = catalog["apcorr_base_CompensatedTophatFlux_12_used"]
+        ratio_used = np.mean(
+            catalog["slot_CalibFlux_instFlux"][used]/catalog[self.ap_name + "_instFlux"][used]
+        )
+        self.assertFloatsAlmostEqual(ratio_used, 1.0, rtol=1e-10)
```
Comment (taranu) [inline tests/test_normalizedCalibrationFlux.py:143]: Is this necessary given the full set check above? It can't hurt, I suppose.

File: tests/test_normalizedCalibrationFlux.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            )
+            self.assertFloatsAlmostEqual(ratio_used, 1.0, rtol=1e-10)
+
+    def testNormalizedCalibrationFluxTooFew(self):
+        np.random.seed(12345)
+        norm_task = self._make_task()
+        catalog = self._make_catalog(norm_task.schema)
+
+        flags = np.ones(len(catalog), dtype=bool)
+        flags[0] = False
```
Comment (taranu) [inline tests/test_normalizedCalibrationFlux.py:186]: Just one valid star is enough to run?

File: tests/test_normalizedCalibrationFlux.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+            norm_task2.config.normalized_calibflux_name,
+        )
+
+        # Check that the final normalized catalog values are the same.
+        self.assertFloatsAlmostEqual(
+            catalog_run2["slot_CalibFlux_instFlux"],
+            catalog_run1["slot_CalibFlux_instFlux"],
+        )
+
+    def testNormalizedCalibrationFluxApplyOnlyFail(self):
```
Comment (taranu) [inline tests/test_normalizedCalibrationFlux.py:236]: Another nitpick but is it conventional to include the name of the test case class in the test member methods?

**Review by taranu — COMMENTED** [review]
Looks good; comments are mostly requests for clarification and/or nitpicks which you can deal with as you see fit.

File: python/lsst/meas/algorithms/measureApCorr.py [hunk 4]
```diff
@@ -140,6 +140,11 @@ class MeasureApCorrConfig(lsst.pex.config.Config):
         dtype=float,
         default=4.0,
     )
+    doMedianShift = lsst.pex.config.Field(
+        doc="Do final shift to ensure medians match.",
```
[the remark this answers is not in the document]
Comment (erykoff) [reply python/lsst/meas/algorithms/measureApCorr.py:144]: I'm happy to rename if you think that makes more sense, and I think it does.

[the remark this answers is not in the document]
Comment (taranu) [reply python/lsst/meas/algorithms/measureApCorr.py:144]: Sure, go ahead.

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+        if self.config.do_measure_ap_corr:
+            self.makeSubtask(
+                "measure_ap_corr",
+                schema=schema,
+                namesToCorrect=[self.config.raw_calibflux_name],
+            )
+
+        name = self.config.normalized_calibflux_name
+        self.flux_name = name + "_instFlux"
```
[the remark this answers is not in the document]
Comment (erykoff) [reply python/lsst/meas/algorithms/normalizedCalibrationFlux.py:119]: 3 is right at the border if that is useful, and I'm just going to leave it.

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            ap_corr_err_field = ap_corr_map[raw_fluxerr_name]
+
+        corrections = ap_corr_field.evaluate(
+            catalog["slot_Centroid_x"],
+            catalog["slot_Centroid_y"],
+        )
+
+        input_flux_name = raw_flux_name
+        input_fluxerr_name = raw_fluxerr_name
+        input_flag_name = raw_name + "_flag"
```
[the remark this answers is not in the document]
Comment (erykoff) [reply python/lsst/meas/algorithms/normalizedCalibrationFlux.py:199]: I guess I copied from older code and just left it...

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        output_flux_name = norm_name + "_instFlux"
+        output_fluxerr_name = norm_name + "_instFluxErr"
+        output_flag_name = norm_name + "_flag"
+
+        if catalog.isContiguous():
```
[the remark this answers is not in the document]
Comment (erykoff) [reply python/lsst/meas/algorithms/normalizedCalibrationFlux.py:204]: If it's non-contiguous you must loop.

File: tests/test_normalizedCalibrationFlux.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+        """
+
+        Parameters
+        ----------
+        apply_only : `bool`, optional
+            Configure task in apply_only mode?
+
+        Returns
+        -------
+        norm_task : `lsst.meas.algorithms.NormalizedCalibrationFluxTask`
```
[the remark this answers is not in the document]
Comment (erykoff) [reply tests/test_normalizedCalibrationFlux.py:48]: Until that's in the dev guide and actually supported, I'm going to use the regular numpydoc style.

File: tests/test_normalizedCalibrationFlux.py (hunk 1 above)
Comment (erykoff) [reply tests/test_normalizedCalibrationFlux.py:143]: Always good to explicitly confirm expectations.

File: tests/test_normalizedCalibrationFlux.py (hunk 2 above)
Comment (erykoff) [reply tests/test_normalizedCalibrationFlux.py:186]: Should be... it's meant to be as robust as possible.

File: tests/test_normalizedCalibrationFlux.py (hunk 3 above)
Comment (erykoff) [reply tests/test_normalizedCalibrationFlux.py:236]: Are these too wordy?  I tend to get verbose with my test names because it's much better than being too terse, but if you think I should shorten them I can.

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+        self.measure_ap_corr.fitConfig.orderY = 0
+
+        fallback_selector = self.fallback_source_selector["science"]
+        fallback_selector.doFluxLimit = False
+        fallback_selector.doFlags = True
+        fallback_selector.doUnresolved = False
+        fallback_selector.doSignalToNoise = False
+        fallback_selector.doIsolated = False
+        fallback_selector.flags.good = ["calib_psf_used"]
+        fallback_selector.flags.bad = []
```
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/normalizedCalibrationFlux.py:90]: I'm missing some connection to Robert's slide here. I was expecting to see a corresponding field for `R1` and `R2`, for the annulus. Is this implicit in `measure_ap_corr` somehow? 🤔

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+        """Internal method to do the aperture correction measurement.
+
+        This measures the aperture correction with the regular task,
+        and if that fails does a fallback median estimate.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure the normalized calibration flux is measured on.
+            This is only used for the bounding box.
```
Comment (arunkannawadi) [inline python/lsst/meas/algorithms/normalizedCalibrationFlux.py:236]: For an internal method, it seems like just passing the `bbox` directly without the `exposure` should do.

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py (hunk 9 above)
Comment (erykoff) [reply python/lsst/meas/algorithms/normalizedCalibrationFlux.py:90]: That's in `meas_base` which measures a compensated tophat (or compensated gaussian).  This is a task to _normalize_ a generic calibration flux that could be a compensated tophat or compensated gaussian or something else we haven't thought of yet.

Comment (arunkannawadi) [reply python/lsst/meas/algorithms/normalizedCalibrationFlux.py:90]: 🤦🏼 I over prepared for this review then. This all LGTM.

File: python/lsst/meas/algorithms/normalizedCalibrationFlux.py (hunk 10 above)
Comment (arunkannawadi) [reply python/lsst/meas/algorithms/normalizedCalibrationFlux.py:236]: Actually, is this true, or at least limited to this method? It looks like this gets passed on to `measure_ap_corr` and this docstring shouldn't comment what `measure_ap_corr` needs from the `exposure`.
